File: 프로그래머스/2/86052. 빛의 경로 사이클/빛의 경로 사이클.py
def solution(grid):
    R, C = len(grid), len(grid[0])
    dxy = [(-1, 0), (0, 1), (1, 0), (0, -1)]  # 상우하좌
    grf = [[[0,0,0,0] for _ in range(C)] for _ in range(R)]  # 상우하좌


    def solve(x, y, d):
        res = 0
        while not grf[x][y][d]:
            grf[x][y][d] = 1
            x, y = (x+dxy[d][0]) % R, (y+dxy[d][1]) % C
            if grid[x][y] == 'L': d = (d+1) % 4
            elif grid[x][y] == 'R': d = (d-1) % 4
            res += 1
        return res


    ans = []
    for i in range(R):
        for j in range(C):
            for k in range(4):
                if grf[i][j][k] == 0:
                    ans.append(solve(i, j, k))             
    return sorted(ans)

File: 프로그래머스/2/86052. 빛의 경로 사이클/빛의 경로 사이클.js
function solution(grid) {
    const R = grid.length, C = grid[0].length;
    const drc = [[-1,0], [0,1], [1,0], [0,-1]];
    const grf = [];
    for (let r=0; r<R; r++) {
        const row = [];
        for (let c=0; c<C; c++) {
            const cell = [false, false, false, false];
            row.push(cell)
        }
        grf.push(row);
    }
    
    const solve = (r, c, d) => {
        let res = 0;
        while (!grf[r][c][d]) {
            res++;
            grf[r][c][d] = true;
            r = (r + drc[d][0] + R) % R;
            c = (c + drc[d][1] + C) % C;
            if (grid[r][c] === 'L') { d = (d+1)%4; }
            else if (grid[r][c] === 'R') { d = (d+3)%4; }
        }
        return res;
    }
    
    
    const ans = [];
    for (let r=0; r<R; r++) {
        for (let c=0; c<C; c++) {
            for (let d=0; d<4; d++) {
                const cycleLength = solve(r, c, d);
                if (cycleLength > 0) {
                    ans.push(cycleLength);
                }
            }
        }
    }
    return ans.sort((a,b) => a-b)
}